pkg/service/service.go

```diff
@@ -1564,20 +1569,23 @@ func (s *Service) restoreServicesLocked(svcBackendsById map[lb.BackendID]struct{
 func (s *Service) deleteServiceLocked(svc *svcInfo) error {
 	ipv6 := svc.frontend.L3n4Addr.IsIPv6() || svc.svcNatPolicy == lb.SVCNatPolicyNat46
 	obsoleteBackendIDs := s.deleteBackendsFromCacheLocked(svc)
+	headless := svc.svcType == lb.SVCTypeNone
 	scopedLog := log.WithFields(logrus.Fields{
 		logfields.ServiceID: svc.frontend.ID,
 		logfields.ServiceIP: svc.frontend.L3n4Addr,
 		logfields.Backends:  svc.backends,
 	})
 	scopedLog.Debug("Deleting service")
 
-	if err := s.lbmap.DeleteService(svc.frontend, len(svc.backends),
-		svc.useMaglev(), svc.svcNatPolicy); err != nil {
-		return err
+	if !headless {
+		if err := s.lbmap.DeleteService(svc.frontend, len(svc.backends),
+			svc.useMaglev(), svc.svcNatPolicy); err != nil {
+			return err
+		}
 	}
 
 	// Delete affinity matches
-	if option.Config.EnableSessionAffinity && svc.sessionAffinity {
+	if !headless && option.Config.EnableSessionAffinity && svc.sessionAffinity {
```

Comment:
It seems to me that also the `s.lbmap.UpdateSourceRanges` call below should also guarded by `!headless`, since the corresponding insertion is never performed (mostly for consistency, since I think this should never happen with an headless service).